### PR TITLE
feat: support penalty

### PR DESCRIPTION
**Pull request by JamesBrianD**

### CODE REVIEW

File: python/sgl_jax/srt/layers/sampler.py [hunk 1]
```diff
@@ -76,6 +76,49 @@ def _process_logprob_results(self, operands):
...
+            # Apply min new tokens penalty to stop tokens
+            min_new_tokens_mask = (
+                sampling_metadata.len_output_tokens < sampling_metadata.min_new_tokens
+            )
+            stop_penalty = jnp.where(
+                min_new_tokens_mask.reshape(-1, 1),
+                sampling_metadata.stop_token_penalties,
+                0.0,
+            )
+            logits = logits - stop_penalty
```
[the remark this answers is not in the document]
Comment (JamesBrianD) [reply python/sgl_jax/srt/layers/sampler.py]: fixed

File: python/sgl_jax/srt/sampling/sampling_batch_info.py [hunk 2]
```diff
@@ -240,15 +729,26 @@ def merge_batch(self, other: "SamplingBatchInfo", mesh: Mesh):
...
     bs1: int,
     bs2: int,
     default: float,
-    mesh: mesh_lib.Mesh,
+    mesh: mesh_lib.Mesh = None,
 ):
+    del mesh  # Parameter not used in current implementation
     """Merge two bias array for batch merging.
 
     Args:
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/sampling_batch_info.py]: Please add penalty tests under test/srt/test_features.py.

It is better to test the cartesian combination of penalties.

Other: lack of repetition penalty.

File: python/sgl_jax/srt/sampling/sampling_batch_info.py [hunk 3]
```diff
@@ -122,6 +154,368 @@ def from_model_worker_batch(
...
+
+                    frequency_penalties_device = device_array(
+                        padded_freq_penalties,
+                        sharding=(
+                            NamedSharding(mesh, PartitionSpec())
+                            if jax.process_count() == 1
+                            else None
+                        ),
+                    )
+                    cumulated_frequency_penalties_device = device_array(
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/sampling_batch_info.py]: please use `device_array((padded_freq_penalties, padded_cumulated_frequency_penalties), sharding=())` to reduce the repeated codes.

File: python/sgl_jax/srt/sampling/sampling_batch_info.py [hunk 4]
```diff
@@ -122,6 +154,368 @@ def from_model_worker_batch(
...
+
+                    presence_penalties_device = device_array(
+                        padded_pres_penalties,
+                        sharding=(
+                            NamedSharding(mesh, PartitionSpec())
+                            if jax.process_count() == 1
+                            else None
+                        ),
+                    )
+                    cumulated_presence_penalties_device = device_array(
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/sampling_batch_info.py]: please use `device_array((padded_freq_penalties, padded_cumulated_frequency_penalties), sharding=())` to reduce the repeated codes.

File: python/sgl_jax/srt/sampling/sampling_batch_info.py [hunk 5]
```diff
@@ -122,6 +154,368 @@ def from_model_worker_batch(
...
+                    )
+                    stop_token_penalties_device = device_array(
+                        padded_stop_token_penalties,
+                        sharding=(
+                            NamedSharding(mesh, PartitionSpec())
+                            if jax.process_count() == 1
+                            else None
+                        ),
+                    )
+                    len_output_tokens_device = device_array(
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/sampling_batch_info.py]: please use `device_array((padded_freq_penalties, padded_cumulated_frequency_penalties), sharding=())` to reduce the repeated codes.

File: python/sgl_jax/srt/sampling/sampling_batch_info.py [hunk 6]
```diff
@@ -122,6 +154,368 @@ def from_model_worker_batch(
...
+        # Linear penalty tensor (for overlap mode)
+        linear_penalty_device = device_array(
+            jnp.ones((batch_size, vocab_size), dtype=jnp.float32) * 0.2,
+            sharding=(
+                NamedSharding(mesh, PartitionSpec())
+                if jax.process_count() == 1
+                else None
+            ),
+        )
+
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/sampling_batch_info.py:298]: please put these array together in device_array to reduce repeated codes.

File: python/sgl_jax/srt/sampling/sampling_batch_info.py [hunk 7]
```diff
@@ -202,16 +679,27 @@ def from_schedule_batch(cls, batch: ScheduleBatch, vocab_size: int):
...
 
-    def apply_logits_bias(self, logits: jax.Array):
-        return logits
+    def update_penalties(self):
+        if self.penalizer_orchestrator.is_required:
+            self.linear_penalty = np.zeros(
+                (len(self.temperatures), self.vocab_size),
+                dtype=np.float32,
+            )
+            self.penalizer_orchestrator.apply(self.linear_penalty)
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/sampling_batch_info.py]: no variable is set for apply result, like this:

self.linear_penalty = self.penalizer_orchestrator.apply(self.linear_penalty) or just use in-place modification for numpy array

File: python/sgl_jax/srt/managers/tp_worker_overlap_thread.py [hunk 8]
```diff
@@ -160,6 +160,7 @@ def forward_batch_generation(
     ) -> Tuple[None, jax.Array, int]:
         # Create a new copy of sampling_info because it will be updated in-place by the scheduler for the next batch.
         sampling_info = model_worker_batch.sampling_info
+        sampling_info.update_penalties()
         model_worker_batch.sampling_info = self.cur_sampling_info = dataclasses.replace(
             sampling_info,
             sampling_info_done=threading.Event(),
```
Comment (aolemila) [inline python/sgl_jax/srt/managers/tp_worker_overlap_thread.py:166]: lack of the following codes:
            penalizer_orchestrator=None,

compared with sglang

File: python/sgl_jax/srt/layers/sampler.py [hunk 9]
```diff
@@ -76,6 +76,119 @@ def _process_logprob_results(self, operands):
 
         return None
 
+    def _apply_linear_penalty(self, operands):
+        """Apply linear penalty branch (overlap mode)"""
+        logits, sampling_metadata = operands
+        penalty = sampling_metadata.linear_penalty
+        return logits - (
```
Comment (aolemila) [inline python/sgl_jax/srt/layers/sampler.py]: It looks like the SGLang uses add.

```python3
    def apply_logits_bias(self, logits: torch.Tensor):
        if self.linear_penalty is not None:
            # Used in the overlap mode
            logits.add_(self.linear_penalty)
```

File: python/sgl_jax/srt/layers/sampler.py [hunk 10]
```diff
@@ -76,6 +76,119 @@ def _process_logprob_results(self, operands):
...
+
+        penalty_operands = (logits, sampling_metadata)
+
+        # Use lax.cond to choose between linear penalty (overlap mode)
+        # and individual penalties (non-overlap mode)
+        return lax.cond(
+            sampling_metadata.linear_penalty is not None,
+            self._apply_linear_penalty,
+            self._apply_individual_penalties,
+            penalty_operands,
```
Comment (aolemila) [inline python/sgl_jax/srt/layers/sampler.py]: It is different with

```
    def apply_logits_bias(self, logits: torch.Tensor):
        if self.linear_penalty is not None:
            # Used in the overlap mode
            logits.add_(self.linear_penalty)

        if self.penalizer_orchestrator and self.penalizer_orchestrator.is_required:
            # Used in the non-overlap mode
            self.penalizer_orchestrator.apply(logits)
```

After apply linear_penalty, the function does not return immediately.

File: python/sgl_jax/srt/managers/tp_worker.py [hunk 11]
```diff
@@ -191,6 +191,7 @@ def normalize_token_paddings(self):
     def run_precompile(self, future_token_ids_map=None):
         self.precompile_extend(future_token_ids_map)
         self.precompile_decode(future_token_ids_map)
+        self.precompile_penalties(future_token_ids_map)
```
Comment (aolemila) [inline python/sgl_jax/srt/managers/tp_worker.py:196]: May be we could put precompile_penalties logic into precompile_extend and precompile_decode after you merge in to main and rebase feat/integrate-tunix with main.

File: python/sgl_jax/srt/layers/sampler.py (hunk 9 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/layers/sampler.py]: ok, I'll fix it.

File: python/sgl_jax/srt/sampling/sampling_batch_info.py (hunk 2 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/sampling/sampling_batch_info.py]: Already added

File: python/sgl_jax/srt/sampling/sampling_batch_info.py (hunk 7 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/sampling/sampling_batch_info.py]: Fixed it. Used in place of updating `np array`.

File: python/sgl_jax/srt/managers/tp_worker_overlap_thread.py (hunk 8 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/managers/tp_worker_overlap_thread.py:166]: What a good reminder.

File: python/sgl_jax/srt/layers/sampler.py (hunk 10 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/layers/sampler.py]: There is no problem here. In the sglang implementation, add linear_penalty in overlap mode instead of applying it in overlap mode.

File: python/sgl_jax/srt/sampling/penaltylib/frequency_penalty.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        self.token_frequencies = np.zeros(
+            (len(self.orchestrator.reqs()), self.orchestrator.vocab_size),
+            dtype=np.int32,
+        )
+
+    def _cumulate_output_tokens(self, output_ids: np.ndarray):
+        batch_indices = np.arange(len(output_ids))
+        self.token_frequencies[batch_indices, output_ids] += 1
+
+    def _apply(self, logits: np.ndarray) -> np.ndarray:
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/penaltylib/frequency_penalty.py]: `_apply` is not used, please remove it for every penalizer.

File: python/sgl_jax/srt/sampling/penaltylib/min_new_tokens.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        mask = self.len_output_tokens < self.min_new_tokens
+
+        # Create stop token penalties on-demand
+        stop_token_penalties = self._create_stop_token_penalties()
+
+        mask_expanded = np.broadcast_to(mask, stop_token_penalties.shape)
+        penalty_values = np.where(mask_expanded, stop_token_penalties, 0.0)
+
+        # Return negative values since orchestrator will subtract (but min_new_tokens should be added)
+        return -penalty_values
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/penaltylib/min_new_tokens.py]: It looks like that the returned value is positive. Please check it and update the annotation.

File: python/sgl_jax/srt/sampling/sampling_batch_info.py [hunk 14]
```diff
@@ -140,6 +175,175 @@ def from_model_worker_batch(
...
+                )
+                padded_min_new_tokens = np.concat(
+                    [
+                        original_min_new_tokens,
+                        np.array([0] * pad_size, dtype=original_min_new_tokens.dtype),
+                    ]
+                )
+
+                original_stop_token_penalties = (
+                    min_tokens_penalizer.stop_token_penalties
```
Comment (aolemila) [inline python/sgl_jax/srt/sampling/sampling_batch_info.py]: It looks like min_tokens_penalizer does not have this filed.

File: test/srt/test_features.py [hunk 15]
```diff
@@ -493,6 +498,221 @@ def test_logprobs(self):
...
+        self.assertIn("text", resp_no_penalty)
+
+        # With frequency penalty, the output should be different (less repetitive)
+        # This is a behavioral test - the penalty should reduce repetition
+        output_with_penalty = resp_with_penalty["text"]
+        output_no_penalty = resp_no_penalty["text"]
+
+        # Basic validation that the responses are generated
+        self.assertIsInstance(output_with_penalty, str)
+        self.assertIsInstance(output_no_penalty, str)
```
Comment (aolemila) [inline test/srt/test_features.py]: please add cache_miss check for every penalty.

refer to https://github.com/sgl-project/sglang-jax/blob/4fa24afba321579fb17cc813883f8ea9614b4c36/test/srt/test_features.py#L555

Comment (aolemila) [reply test/srt/test_features.py]: It looks like codes do not check differences between two responses. Please check the follow two penalties.

File: test/srt/test_features.py [hunk 16]
```diff
@@ -493,6 +498,221 @@ def test_logprobs(self):
...
+        output_combined = resp_combined["text"]
+        output_no_penalty = resp_no_penalty["text"]
+
+        self.assertIsInstance(output_combined, str)
+        self.assertIsInstance(output_no_penalty, str)
+
+        # Both should generate some output beyond the input
+        self.assertGreater(len(output_combined), len(args.text))
+        self.assertGreater(len(output_no_penalty), len(args.text))
+
```
Comment (aolemila) [inline test/srt/test_features.py:794]: please add another two cases:

1. scheduler normal mode
2. combination of three penalties: please check cache_miss

File: python/sgl_jax/srt/sampling/penaltylib/frequency_penalty.py (hunk 12 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/sampling/penaltylib/frequency_penalty.py]: DONE
